crates/ra_ide_api/src/references.rs

```diff
@@ -52,161 +58,88 @@ pub(crate) fn find_all_refs(
-    new_name: &str,
-) -> Option<SourceChange> {
-    let mut source_file_edits = Vec::new();
-    let mut file_system_edits = Vec::new();
-    let module_src = hir::Source { file_id: position.file_id.into(), ast: ast_module.clone() };
-    if let Some(module) = hir::Module::from_declaration(db, module_src) {
-        let src = module.definition_source(db);
-        let file_id = src.file_id.original_file(db);
-        match src.ast {
-            ModuleSource::SourceFile(..) => {
-                let mod_path: RelativePathBuf = db.file_relative_path(file_id);
-                // mod is defined in path/to/dir/mod.rs
-                let dst_path = if mod_path.file_stem() == Some("mod") {
-                    mod_path
-                        .parent()
-                        .and_then(|p| p.parent())
-                        .or_else(|| Some(RelativePath::new("")))
-                        .map(|p| p.join(new_name).join("mod.rs"))
-                } else {
-                    Some(mod_path.with_file_name(new_name).with_extension("rs"))
-                };
-                if let Some(path) = dst_path {
-                    let move_file = FileSystemEdit::MoveFile {
-                        src: file_id,
-                        dst_source_root: db.file_source_root(position.file_id),
-                        dst_path: path,
-                    };
-                    file_system_edits.push(move_file);
-                }
-            }
-            ModuleSource::Module(..) => {}
+fn process_definition(db: &RootDatabase, def: NameDefinition, name: String) -> Vec<FileRange> {
+    let pat = name.as_str();
+    let scope = def.scope(db).scope;
+    let mut refs = vec![];
+
+    let is_match = |file_id: FileId, name_ref: &ast::NameRef| -> bool {
+        let classified = classify_name_ref(db, file_id, &name_ref);
+        if let Some(d) = classified {
+            d == def
+        } else {
+            false
         }
-    }
-
-    let edit = SourceFileEdit {
-        file_id: position.file_id,
-        edit: {
-            let mut builder = ra_text_edit::TextEditBuilder::default();
-            builder.replace(ast_name.syntax().text_range(), new_name.into());
-            builder.finish()
-        },
     };
-    source_file_edits.push(edit);
 
-    Some(SourceChange::from_edits("rename", source_file_edits, file_system_edits))
-}
-
-fn rename_reference(
-    db: &RootDatabase,
-    position: FilePosition,
-    new_name: &str,
-) -> Option<RangeInfo<SourceChange>> {
-    let RangeInfo { range, info: refs } = find_all_refs(db, position)?;
-
-    let edit = refs
-        .into_iter()
-        .map(|range| source_edit_from_file_id_range(range.file_id, range.range, new_name))
-        .collect::<Vec<_>>();
-
-    if edit.is_empty() {
-        return None;
+    for (file_id, text_range) in scope {
+        let text = db.file_text(file_id);
+        let parse = SourceFile::parse(&text);
+        let syntax = parse.tree().syntax().clone();
```

Comment:
It's better to parse file only when we found a hit. Most files will not contain hits, and we can save time by not parsing them! This actually might be a good use-case for once_cell::unsync::Lazy